File: C#/20. Valid Parentheses.cs
public class Solution
{
    public bool IsValid(string s)
    {
        Stack<char> stack = new Stack<char>();
        foreach (char chs in s)
        {
            if (chs == '(' || chs == '{' || chs == '[')
            {
                stack.Push(chs);
            }
            if (chs == ')')
            {
                if (stack.Count == 0)
                    return false;
                char top = stack.Peek();
                if (top == '(')
                    stack.Pop();
                else return false;
            }
            if (chs == '}')
            {
                if (stack.Count == 0)
                    return false;
                char top = stack.Peek();
                if (top == '{')
                    stack.Pop();
                else return false;
            }
            if (chs == ']')
            {
                if (stack.Count == 0)
                    return false;
                char top = stack.Peek();
                if (top == '[')
                    stack.Pop();
                else return false;
            }
        }
        return stack.Count == 0;
    }
}

File: python/20. Valid Parentheses..py
# -*- coding: utf-8 -*-

"""
@Author: xiezizhe
@Date: 26/2/2020 上午12:57
"""


class Solution:
    def pair(self, a, b):
        return (a == '(' and b == ')') or (a == '[' and b == ']') or (a == '{' and b == '}')

    def isValid(self, s: str) -> bool:
        if s is None or len(s) == 0:
            return True
        stack = []
        for c in s:
            if len(stack) == 0:
                stack.append(c)
                continue
            if self.pair(stack[-1], c):
                stack.pop()
            else:
                stack.append(c)
        return len(stack) == 0
